src/java.base/share/classes/jdk/internal/foreign/abi/BufferStack.java

```diff
@@ -0,0 +1,55 @@
+package jdk.internal.foreign.abi;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class BufferStack {
+    private final MemorySegment backingSegment;
+    private final ReentrantLock lock = new ReentrantLock();
+    private long offset = 0;
+
+    public BufferStack(MemorySegment backingSegment) {
+        this.backingSegment = backingSegment;
+    }
+
+    public Arena reserve(long size) {
+        if (!lock.tryLock()) {
```

Comment:
I believe we only need to lock if `reserve` is called from a virtual thread. So there's more cases where the locking can be avoided :-)